README.md
# CoD4x Server [![Build Status](https://travis-ci.org/callofduty4x/CoD4x_Server.svg?branch=master)](https://travis-ci.org/callofduty4x/CoD4x_Server) [![Build status](https://ci.appveyor.com/api/projects/status/oalrrpp4j49iqdgy?svg=true&branch=master)](https://ci.appveyor.com/project/D4edalus/cod4x-server)
CoD4x is a modification of the Call of Duty 4 - Modern Warfare server. It fixes several bugs in the original binaries and allows developers to extend server functionality with additional variables and plugins. When using the CoD4x server, the clients invoke  installation of the proprietary CoD4x client to players joining the server using the CoD4x servers, which fixes several known base game bugs in the client, and in combination with the server allows for extra features.

## The most prominent features are:
* Administration commands
* Measurements against hackers
* Reliable player identification
* Pbss like screenshot feature
* Automated client update, no manual installation needed
* Backwards compatibility to 1.7 and 1.7a servers
* A new [masterserver](http://cod4master.cod4x.me/), for when the official masterserver is down

The CoD4x server can run on Windows and Linux. 
The CoD4x client update is only available for windows.

## Setting up a Call of Duty 4 server with Cod4x 1.8
Download binaries: [Linux](https://cod4x.me/downloads/cod4x_server-linux.zip) [Windows](https://cod4x.me/downloads/cod4x_server-windows.zip)

You also require the base game to run a server. Copy every .iwd file in `cod4directory/main/` to `serverdirectory/main/`.
Also copy everything inside `cod4directory/zone` to `serverdirectory/zone`.

Now you can run the server with `./cod4x18_dedrun +map mp_killhouse`. If you are running a local server on Windows use `cod4x18_dedrun.exe +map mp_killhouse +set dedicated 2 +set net_ip 127.0.0.1`. Join the server with your client via the console (`^`) by typing `/connect 127.0.0.1` (if hosted locally), and see if you can join the server.

Hint: you probably want to run the server on a separate user. Please don't run the server (any server) as root. That would be a major security threat.

A more detailed server tutorial is available on [our wiki](https://github.com/callofduty4x/CoD4x_Server/wiki/Server-setup).
[Also read about new banlists here](https://github.com/callofduty4x/CoD4x_Server/wiki/Banlists-in-version-15.9--and-other-changes)

## Compiling on Linux
To compile CoD4x from source you require the following tools:

- NASM
- paxctl
- gcc
- make

Debian/Ubuntu 32-bit:
```
sudo apt install nasm paxctl build-essential
```

Debian/Ubuntu 64-bit:
```
sudo dpkg --add-architecture i386
sudo apt-get update
sudo apt-get install nasm:i386 build-essential gcc-multilib g++-multilib
```

openSUSE 32-bit: 
```
sudo zypper install nasm gcc-32bit gcc-c++-32bit
```
 
Arch Linux 32-bit:
```
yaourt -S nasm paxctl gcc-multilib make
```

Now compile the server by running `make`.

If compilation was successful the binary will be placed in the `/bin/` folder.

## Compiling on Windows
To compile CoD4x from source you require the following tools:

- [NASM](http://www.nasm.us/pub/nasm/releasebuilds/?C=M;O=D)
- [MinGW](http://www.mingw.org/): 'gcc' and 'mingw32-make'

Now compile the server by running `mingw32-make`.

If compilation was successful the binary will be placed in the `/bin/` folder.

## Contributing
CoD4x is licensed under the AGPL3 license. We welcome anybody to fork this project and submit a pull request.

Plugins can be written in C/C++ and we also provide language bindings for D. The `/plugins` directory contains some example plugins. You can contribute to the project by developing plugins and create a pull request for them and/or uploading and promoting them on the [forums](https://cod4x.me/forum/forum-17.html).

If you want to contribute to the core project check the issue tracker for todos. We will try our best to keep the issue tracker filled with new bits.
If you would like to work on a completely new feature, we would appreciate if you contact us first on the forums or on Github to discuss the idea.

If you're not a programmer but still want to help, you can help by testing and reporting bugs, but also by writing documentation. Please submit your bug reports to the Github issue tracker.

## Usage conditions for server hosters
Aside from agreeing to the license, by making any use of CoD4x18 server you agree to the following:

1. You make content which is connected to your CoD4x18 Server available to the developers on request. For example if you run a mod, you have to make everything available that is required to run another server just like your own. Think of a complete mod.ff, .iwds, plugins, database handlers, etc.

2. The developers reserve the right to reuse your content as long as it is not used commercially. You have a right for your name/clan/website getting mentioned if this is going to happen.
They can also use it on their own servers.

3. Maps you have installed on a server have to be either available on the internet already, or be made available to the community at the [CoD4x forums](https://cod4x.me/index.php?/forums/), with all required assets, like scripts, within 20 weeks of installation. You have to annouce your map on the [forums](https://cod4x.me/index.php?/forums/) on the same day you have installed it to gain the 20 weeks grace period. Not announced maps will have to be made available within 1 week.

4. Plugins have to be made available as sourcecode so the user can interact with it

Server's IPs violating these conditions can get permanently disabled.

These conditions have been established to keep user created content open to everyone, and also to value the work on CoD4x.

## Everything else
Please check out the [forums](https://cod4x.me) and [our wiki](https://github.com/D4edalus/CoD4x_Server/wiki).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

Versions in this file correspond to git tags.

## Unreleased
### Changes
#### Makefiles
Now you can use "sv_disableChat" dvar to disable chat messages on server (commands "$..." still available).
incremental builds are now possible. `make -f makefile2` runs the new build scripts. appveyor and travis continuous integration services have been updated to the new buildprocess.

## 17.3
Default fs_homepath under Windows is now the current working directory - before default fs_homepath was fs_basepath

### Fixes
Masterserver announcing works now with more than 1 address - was a bug before 

Small zone overflow [#148](https://github.com/callofduty4x/CoD4x_Server/issues/148)

### Changes
#### Autoupdate [#150](https://github.com/callofduty4x/CoD4x_Server/issues/150)
Autoupdate will check for updates earlier in startup

Updatefiles will now only written to these locations: 
-    The cod4x18_dedrun,
-    Windows: Executable-directory
-    Linux: Current working directory,
-    fs_homepath

User compiled cod4x server will only receive updates on major version increases

Network init happens now early to allow early run of autoupdater

#### Note: For proper operation of Autoupdate every gameserver needs to run on a dedicated fs_homepath

## 17.0
Note: 17.0 is a major update. Servers and Clients will be forced to upgrade.

xbase.iwd must be updated!

### Fixed
- Demo stopping on fast_restart
- Memoryleaks
- LAN broadcasting

### Changed
- Masterserver now in sv_masterservers
- Player stats now encrypted
- Colored playersnames are filtered
- homepath overrides basepath for file discovery

### Added
- Scripting Guide
- Support for higher xasset limits

## 16.3
### Fixed
- Masterserver Bufferoverflow

### Added
- Builds with Makefiles
- Plugin_SetStat, Plugin_GetStat

### Changed
- Support for binding multiple IP addresses has been dropped


### Fixed

## 16.2
### Added
- ScreenshotArrived Plugin Callback
- Script functions to check player keys
- Scr_GetFunc
- Screenshot Uploader Plugin
- IsCvarDefined Script Function
- Array Manipulation Script Functions
- Script functions to control bots
- Menu File Extraction

### Changed
- Standard GUID format changed to 64bit IDs
- Removed bansystem from Server Core
- Http Api exposed to plugins ( blocking / non-blocking )

## Before ...

July 7 2013:

Code_CallbackPlayerSay*: Added 2nd argument (to entity)

August 11 2013
Code_CallbackPlayerSay*: 2nd argument = boolean teamchat -> false: toall true: onlyteam


Hmm I was too lazy to write anything...

February 3 2014:

CodeCallback_PlayerSayCmd is removed.
CodeCallback_PlayerSayAll is renamed to CodeCallback_PlayerSay.
CodeCallback_ScriptCommand added as well as script function: addscriptcommand()

Fixed an issue that cheatprotected cvars got no reset on level startup which caused strange effects on Deathrun mod with bg_falldamage cvars.

June 22 2014:
Removed chat trap for script when a command gets invoked.
Use CodeCallback_ScriptCommand and addscriptcommand() instead  (see scriptcommand_example.txt)

Added securemode which should prevent breaking out of cod4 server into shell.


June 29 2014:
Removed the non working recusion callback function for player's chat. Use plugins instead.

July 27 2014:
exec() script function now returns the result as string truncated - Maybe this functionality getting removed later again if issues arise

September 09 2014:
exec is now execex
exec works like before July 27 2014

September 12 2016:
After a 2 years its still good to write about changes in server code ^^
 - 'SetText' script command now safe to use with dynamically created strings like player names, number increments and etc.
 - 'ClearAllTextAfterHudElem' script command now can be used only in developer mode because of its bugs.

September 13 2016:
 + Added 'Scr_AddArrayKey' script and plugin function. It allows you to create script arrays with string keys instead of numeric. Just make sure you passed value returned from 'Scr_AllocString' (call this one on initialization).

September 14 2016:
 + Added 'isArray' script function. (http://zeroy.com/script/variables/isarray.htm)



dis/md5.c
/*
 * This code implements the MD5 message-digest algorithm.
 * The algorithm is due to Ron Rivest.  This code was
 * written by Colin Plumb in 1993, no copyright is claimed.
 * This code is in the public domain; do with it what you wish.
 *
 * Equivalent code is available from RSA Data Security, Inc.
 * This code has been tested against that, and is equivalent,
 * except that you don't need to include two pages of legalese
 * with every copy.
 *
 * To compute the message digest of a chunk of bytes, declare an
 * MD5Context structure, pass it to MD5Init, call MD5Update as
 * needed on buffers full of bytes, and then call MD5Final, which
 * will fill a supplied 16-byte array with the digest.
 */
//#include "q_shared.h"
//#include "qcommon.h"

typedef struct MD5Context {
	uint32_t buf[4];
	uint32_t bits[2];
	unsigned char in[64];
} MD5_CTX;

#ifndef Q3_BIG_ENDIAN
	#define byteReverse(buf, len)	/* Nothing */
#else
	static void byteReverse(unsigned char *buf, unsigned longs);

	/*
	 * Note: this code is harmless on little-endian machines.
	 */
	static void byteReverse(unsigned char *buf, unsigned longs)
	{
	    uint32_t t;
	    do {
		t = (uint32_t)
			((unsigned) buf[3] << 8 | buf[2]) << 16 |
			((unsigned) buf[1] << 8 | buf[0]);
		*(uint32_t *) buf = t;
		buf += 4;
	    } while (--longs);
	}
#endif // Q3_BIG_ENDIAN

/*
 * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
 * initialization constants.
 */
static void MD5Init(struct MD5Context *ctx)
{
    ctx->buf[0] = 0x67452301;
    ctx->buf[1] = 0xefcdab89;
    ctx->buf[2] = 0x98badcfe;
    ctx->buf[3] = 0x10325476;

    ctx->bits[0] = 0;
    ctx->bits[1] = 0;
}
/* The four core functions - F1 is optimized somewhat */

/* #define F1(x, y, z) (x & y | ~x & z) */
#define F1(x, y, z) (z ^ (x & (y ^ z)))
#define F2(x, y, z) F1(z, x, y)
#define F3(x, y, z) (x ^ y ^ z)
#define F4(x, y, z) (y ^ (x | ~z))

/* This is the central step in the MD5 algorithm. */
#define MD5STEP(f, w, x, y, z, data, s) \
	( w += f(x, y, z) + data,  w = w<<s | w>>(32-s),  w += x )

/*
 * The core of the MD5 algorithm, this alters an existing MD5 hash to
 * reflect the addition of 16 longwords of new data.  MD5Update blocks
 * the data and converts bytes into longwords for this routine.
 */
static void MD5Transform(uint32_t buf[4],
	uint32_t const in[16])
{
    register uint32_t a, b, c, d;

    a = buf[0];
    b = buf[1];
    c = buf[2];
    d = buf[3];

    MD5STEP(F1, a, b, c, d, in[0] + 0xd76aa478, 7);
    MD5STEP(F1, d, a, b, c, in[1] + 0xe8c7b756, 12);
    MD5STEP(F1, c, d, a, b, in[2] + 0x242070db, 17);
    MD5STEP(F1, b, c, d, a, in[3] + 0xc1bdceee, 22);
    MD5STEP(F1, a, b, c, d, in[4] + 0xf57c0faf, 7);
    MD5STEP(F1, d, a, b, c, in[5] + 0x4787c62a, 12);
    MD5STEP(F1, c, d, a, b, in[6] + 0xa8304613, 17);
    MD5STEP(F1, b, c, d, a, in[7] + 0xfd469501, 22);
    MD5STEP(F1, a, b, c, d, in[8] + 0x698098d8, 7);
    MD5STEP(F1, d, a, b, c, in[9] + 0x8b44f7af, 12);
    MD5STEP(F1, c, d, a, b, in[10] + 0xffff5bb1, 17);
    MD5STEP(F1, b, c, d, a, in[11] + 0x895cd7be, 22);
    MD5STEP(F1, a, b, c, d, in[12] + 0x6b901122, 7);
    MD5STEP(F1, d, a, b, c, in[13] + 0xfd987193, 12);
    MD5STEP(F1, c, d, a, b, in[14] + 0xa679438e, 17);
    MD5STEP(F1, b, c, d, a, in[15] + 0x49b40821, 22);

    MD5STEP(F2, a, b, c, d, in[1] + 0xf61e2562, 5);
    MD5STEP(F2, d, a, b, c, in[6] + 0xc040b340, 9);
    MD5STEP(F2, c, d, a, b, in[11] + 0x265e5a51, 14);
    MD5STEP(F2, b, c, d, a, in[0] + 0xe9b6c7aa, 20);
    MD5STEP(F2, a, b, c, d, in[5] + 0xd62f105d, 5);
    MD5STEP(F2, d, a, b, c, in[10] + 0x02441453, 9);
    MD5STEP(F2, c, d, a, b, in[15] + 0xd8a1e681, 14);
    MD5STEP(F2, b, c, d, a, in[4] + 0xe7d3fbc8, 20);
    MD5STEP(F2, a, b, c, d, in[9] + 0x21e1cde6, 5);
    MD5STEP(F2, d, a, b, c, in[14] + 0xc33707d6, 9);
    MD5STEP(F2, c, d, a, b, in[3] + 0xf4d50d87, 14);
    MD5STEP(F2, b, c, d, a, in[8] + 0x455a14ed, 20);
    MD5STEP(F2, a, b, c, d, in[13] + 0xa9e3e905, 5);
    MD5STEP(F2, d, a, b, c, in[2] + 0xfcefa3f8, 9);
    MD5STEP(F2, c, d, a, b, in[7] + 0x676f02d9, 14);
    MD5STEP(F2, b, c, d, a, in[12] + 0x8d2a4c8a, 20);

    MD5STEP(F3, a, b, c, d, in[5] + 0xfffa3942, 4);
    MD5STEP(F3, d, a, b, c, in[8] + 0x8771f681, 11);
    MD5STEP(F3, c, d, a, b, in[11] + 0x6d9d6122, 16);
    MD5STEP(F3, b, c, d, a, in[14] + 0xfde5380c, 23);
    MD5STEP(F3, a, b, c, d, in[1] + 0xa4beea44, 4);
    MD5STEP(F3, d, a, b, c, in[4] + 0x4bdecfa9, 11);
    MD5STEP(F3, c, d, a, b, in[7] + 0xf6bb4b60, 16);
    MD5STEP(F3, b, c, d, a, in[10] + 0xbebfbc70, 23);
    MD5STEP(F3, a, b, c, d, in[13] + 0x289b7ec6, 4);
    MD5STEP(F3, d, a, b, c, in[0] + 0xeaa127fa, 11);
    MD5STEP(F3, c, d, a, b, in[3] + 0xd4ef3085, 16);
    MD5STEP(F3, b, c, d, a, in[6] + 0x04881d05, 23);
    MD5STEP(F3, a, b, c, d, in[9] + 0xd9d4d039, 4);
    MD5STEP(F3, d, a, b, c, in[12] + 0xe6db99e5, 11);
    MD5STEP(F3, c, d, a, b, in[15] + 0x1fa27cf8, 16);
    MD5STEP(F3, b, c, d, a, in[2] + 0xc4ac5665, 23);

    MD5STEP(F4, a, b, c, d, in[0] + 0xf4292244, 6);
    MD5STEP(F4, d, a, b, c, in[7] + 0x432aff97, 10);
    MD5STEP(F4, c, d, a, b, in[14] + 0xab9423a7, 15);
    MD5STEP(F4, b, c, d, a, in[5] + 0xfc93a039, 21);
    MD5STEP(F4, a, b, c, d, in[12] + 0x655b59c3, 6);
    MD5STEP(F4, d, a, b, c, in[3] + 0x8f0ccc92, 10);
    MD5STEP(F4, c, d, a, b, in[10] + 0xffeff47d, 15);
    MD5STEP(F4, b, c, d, a, in[1] + 0x85845dd1, 21);
    MD5STEP(F4, a, b, c, d, in[8] + 0x6fa87e4f, 6);
    MD5STEP(F4, d, a, b, c, in[15] + 0xfe2ce6e0, 10);
    MD5STEP(F4, c, d, a, b, in[6] + 0xa3014314, 15);
    MD5STEP(F4, b, c, d, a, in[13] + 0x4e0811a1, 21);
    MD5STEP(F4, a, b, c, d, in[4] + 0xf7537e82, 6);
    MD5STEP(F4, d, a, b, c, in[11] + 0xbd3af235, 10);
    MD5STEP(F4, c, d, a, b, in[2] + 0x2ad7d2bb, 15);
    MD5STEP(F4, b, c, d, a, in[9] + 0xeb86d391, 21);

    buf[0] += a;
    buf[1] += b;
    buf[2] += c;
    buf[3] += d;
}

/*
 * Update context to reflect the concatenation of another buffer full
 * of bytes.
 */
static void MD5Update(struct MD5Context *ctx, unsigned char const *buf,
	unsigned len)
{
    uint32_t t;

    /* Update bitcount */

    t = ctx->bits[0];
    if ((ctx->bits[0] = t + ((uint32_t) len << 3)) < t)
	ctx->bits[1]++;		/* Carry from low to high */
    ctx->bits[1] += len >> 29;

    t = (t >> 3) & 0x3f;	/* Bytes already in shsInfo->data */

    /* Handle any leading odd-sized chunks */

    if (t) {
	unsigned char *p = (unsigned char *) ctx->in + t;

	t = 64 - t;
	if (len < t) {
	    memcpy(p, buf, len);
	    return;
	}
	memcpy(p, buf, t);
	byteReverse(ctx->in, 16);
	MD5Transform(ctx->buf, (uint32_t *) ctx->in);
	buf += t;
	len -= t;
    }
    /* Process data in 64-byte chunks */

    while (len >= 64) {
	memcpy(ctx->in, buf, 64);
	byteReverse(ctx->in, 16);
	MD5Transform(ctx->buf, (uint32_t *) ctx->in);
	buf += 64;
	len -= 64;
    }

    /* Handle any remaining bytes of data. */

    memcpy(ctx->in, buf, len);
}


/*
 * Final wrapup - pad to 64-byte boundary with the bit pattern 
 * 1 0* (64-bit count of bits processed, MSB-first)
 */
static void MD5Final(struct MD5Context *ctx, unsigned char *digest)
{
    unsigned count;
    unsigned char *p;

    /* Compute number of bytes mod 64 */
    count = (ctx->bits[0] >> 3) & 0x3F;

    /* Set the first char of padding to 0x80.  This is safe since there is
       always at least one byte free */
    p = ctx->in + count;
    *p++ = 0x80;

    /* Bytes of padding needed to make 64 bytes */
    count = 64 - 1 - count;

    /* Pad out to 56 mod 64 */
    if (count < 8) {
	/* Two lots of padding:  Pad the first block to 64 bytes */
	memset(p, 0, count);
	byteReverse(ctx->in, 16);
	MD5Transform(ctx->buf, (uint32_t *) ctx->in);

	/* Now fill the next block with 56 bytes */
	memset(ctx->in, 0, 56);
    } else {
	/* Pad block to 56 bytes */
	memset(p, 0, count - 8);
    }
    byteReverse(ctx->in, 14);

    /* Append length in bits and transform */
    ((uint32_t *) ctx->in)[14] = ctx->bits[0];
    ((uint32_t *) ctx->in)[15] = ctx->bits[1];

    MD5Transform(ctx->buf, (uint32_t *) ctx->in);
    byteReverse((unsigned char *) ctx->buf, 4);
    
    if (digest!=NULL)
	    memcpy(digest, ctx->buf, 16);
    memset(ctx, 0, sizeof(ctx));	/* In case it's sensitive */
}


char *Com_MD5( const char *string)
{
	static char final[33] = {""};
	unsigned char digest[16] = {""}; 
	MD5_CTX md5;
	unsigned char buffer[1024];
	int i;
	int r = strlen(string)+1;

	Q_strncpyz((char*)buffer, string, sizeof(buffer));
	Q_strncpyz( final, "", sizeof( final ) );


	MD5Init(&md5);
	MD5Update(&md5 , buffer, r);
	MD5Final(&md5, digest);

	final[0] = '\0';
	for(i = 0; i < 16; i++) {
		Q_strcat(final, sizeof(final), va("%02x", digest[i]));
	}
	return final;
}